fixed issue caused due to misspellings

thebishalniroula opened the pull request:
Fixed the issue that was being caused due to misspellings.

I found a node package called 'string-similarity' based on [Dice's coefficient](https://en.wikipedia.org/wiki/S%C3%B8rensen%E2%80%93Dice_coefficient)  algorithm to compare the degree of similarity of two strings.

The two strings, one provided by the user and the one on the JSON data are compared against each other. I tried and tested some names of districts with variations in their spelling and the best result was found if the degree of similarity was over 60%.

This is my first contribution to an open-source. So hope I get some feedback on what can be done better.

thebishalniroula committed:
fixed issue caused due to misspellings

ashiishme commented:
@thebishalniroula Thanks for the PR.

Here's few things I would like to notify.

- We will need test to cover this implementation, so it's necessary to write test case for such feature.
- I just tested it and looks like it replace the user input with the best match. It's a good feature but we might not want to do so because in timed mode, user can get advantage because of this implementation ( kind of a hack to guess the name faster ).

Let says, the input was `Rupan` and the best match will be most probably `Rupandehi`. The current implementation replaces `Rupan` to `Rupandehi` and the district is drawn on screen. This gives advantage to users because they don't have to type additional `4 letters: (dehi)`. This applies to other district names as well.

The PR also doesn't fulfill the complete requirement as mentioned in #18. 

We can discuss the implementation on that issue if you want to take over fixing this issue.

thebishalniroula committed:
Implemented the did you mean feature.

thebishalniroula commented:
@ashiishme Okay got it. 

Instead of directly replacing the user input with the probable match, now what it does is show a prompt to the user saying
```
Did you mean? <District name>
```
and when the user clicks over the prompt only then the suggested input is taken.

Please check this out. And tell me if this needs any further changes.

thebishalniroula committed:
User experience improvement on input field

Now the input field gets cleared if the user decides to click on the Did
you mean? prompt

ashiishme commented:
> @ashiishme Okay got it.
> 
> Instead of directly replacing the user input with the probable match, now what it does is show a prompt to the user saying
> 
> ```
> Did you mean? <District name>
> ```
> 
> and when the user clicks over the prompt only then the suggested input is taken.
> 
> Please check this out. And tell me if this needs any further changes.

Can you also include test for this feature? I think we need to write test at least for this case.

aabishkaryal commented:
Please PR into the dev branch

aabishkaryal commented on pages/index.tsx:
```diff
@@ -25,11 +26,42 @@ export default function Home() {
+    } else {
+      setError(`No district named ${userInput}`);
+    }
+  };
```
addDistrictFromHint and addDistrict are very similar maybe you can create a more general function? Just some thoughts.

thebishalniroula replied:
Yup, tried doing that but since the two functions take two different types of parameters the code started getting messy. So decided to addDistrictFromHint implementation in a different function.

aabishkaryal commented on pages/index.tsx:
```diff
@@ -25,11 +26,42 @@ export default function Home() {
+
   const addDistrict = (e: FormEvent<HTMLFormElement>) => {
     e.preventDefault();
+    if (hint != '') {
+      setHint('');
+    }
+    if (error != '') {
+      setError('');
+    }
```
are the if statements really necessary?

aabishkaryal commented on pages/index.tsx:
```diff
@@ -44,7 +76,19 @@ export default function Home() {
         setError('');
       } else {
-        setError(`No district named ${userInput}`);
+        const possibleDistrictIndex = DistrictData.findIndex((d) =>
+          stringSimilarity.compareTwoStrings(
+            d.name.toLowerCase(),
+            userInput.toLowerCase()
+          ) > 0.6
+            ? true
+            : false
+        );
```
you could just return the result of the condition, no need for the ternary operator

aabishkaryal commented on pages/index.tsx:
```diff
@@ -72,6 +116,14 @@ export default function Home() {
         {error != '' && (
           <div className="w-full text-center text-sm text-red-500">{error}</div>
         )}
+        {hint != '' && (
+          <div
+            className="text-black-600 w-full cursor-pointer text-center text-base"
+            onClick={() => addDistrictFromHint(hint)}
+          >
+            Did you mean? <span className="font-semibold">{hint}</span>
+          </div>
+        )}
```
empty string qualify as falsey in js, so the checks are redundant, please fix my checks too in the mean time
